core/utils/exact_match_table.h

```diff
@@ -0,0 +1,443 @@
+    if (fields.size() != num_fields_) {
+      err_ = Format("rule should have %zu fields (has %zu)", num_fields_,
+                    fields.size());
+      return EINVAL;
+    }
+
+    memset(key, 0, sizeof(*key));
+
+    for (size_t i = 0; i < fields.size(); i++) {
+      int field_size = fields_[i].size;
+      int field_pos = fields_[i].pos;
+
+      const std::vector<uint8_t> &f_obj = fields[i];
+
+      if (static_cast<size_t>(field_size) != f_obj.size()) {
+        err_ = Format("rule field %zu should have size %d (has %zu)", i,
+                      field_size, f_obj.size());
+        return EINVAL;
+      }
+
+      memcpy(reinterpret_cast<uint8_t *>(key) + field_pos, f_obj.data(),
+             field_size);
+    }
+
+    return 0;
+  }
+
+  // Helper for public MakeKey functions
+  void __MakeKeys(ExactMatchKey *keys, const void **bufs, size_t n) const {
+    for (size_t i = 0; i < num_fields_; i++) {
+      uint64_t mask = fields_[i].mask;
+      int offset = fields_[i].offset;
+      int pos = fields_[i].pos;
+
+      for (size_t j = 0; j < n; j++) {
+        uint8_t *k = reinterpret_cast<uint8_t *>(keys[j].u64_arr) + pos;
+
+        *(reinterpret_cast<uint64_t *>(k)) =
+            *(reinterpret_cast<const uint64_t *>(
+                reinterpret_cast<const uint8_t *>(bufs[j]) + offset)) &
+            mask;
+      }
+    }
+  }
+
+  // Helper for public AddField functions.
+  // __AddField inserts `field` as the `idx`th field for this table.
+  // If `mt_attr_name` is set, the `offset` field of `field` will be ignored and
+  // the inserted field will use the offset of `mt_attr_name` as reported by the
+  // module `m`.
+  // Returns 0 on success, non-zero errno on failure.
+  int __AddField(const ExactMatchField &field, const std::string &mt_attr_name,
+                 int idx, Module *m = nullptr) {
+    if (idx >= MAX_FIELDS) {
+      err_ = Format("idx %d is not in [0,%d)", idx, MAX_FIELDS);
+      return EINVAL;
+    }
+    ExactMatchField *f = &fields_[idx];
+    f->size = field.size;
+    if (f->size < 1 || f->size > MAX_FIELD_SIZE) {
+      err_ = Format("idx %d: 'size' must be in [1,%d]", idx, MAX_FIELD_SIZE);
+      return EINVAL;
+    }
+
+    if (mt_attr_name.length() > 0) {
+      f->attr_id = m->AddMetadataAttr(mt_attr_name, f->size,
+                                      metadata::Attribute::AccessMode::kRead);
+      if (f->attr_id < 0) {
+        err_ = Format("idx %d: add_metadata_attr() failed", idx);
+        return -f->attr_id;
+      }
+    } else {
+      f->attr_id = -1;
+      f->offset = field.offset;
+      if (f->offset < 0 || f->offset > 1024) {
+        err_ = Format("idx %d: invalid 'offset'", idx);
+        return EINVAL;
+      }
+    }
+
+    int force_be = (f->attr_id < 0);
+
+    if (field.mask == 0) {
+      /* by default all bits are considered */
+      f->mask =
+          (f->size == 8) ? 0xFFFFFFFFFFFFFFFFULL : (1ULL << (f->size * 8)) - 1;
```

Comment:
This one is minor, but I see a couple of cases of "make a bit mask for value" that are like this scattered throughout bess, and they all have the "if 0 then all-1-bits else (1(maybe-ULL) << k) - 1" pattern.  It's short but maybe deserves a dinky little header file for it?

Reply:
Ah, a util like that might fit nicely in `utils/bits.h`. I'll open a separate PR for that